File: StatusBarSpeedometer/app/src/main/java/ch/rmy/android/statusbar_tacho/activities/SettingsActivity.java
package ch.rmy.android.statusbar_tacho.activities;

import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.CompoundButton;
import android.widget.CompoundButton.OnCheckedChangeListener;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.ToggleButton;

import java.util.ArrayList;
import java.util.List;

import butterknife.Bind;
import ch.rmy.android.statusbar_tacho.R;
import ch.rmy.android.statusbar_tacho.services.TachoService;
import ch.rmy.android.statusbar_tacho.units.Unit;
import ch.rmy.android.statusbar_tacho.units.Units;
import ch.rmy.android.statusbar_tacho.utils.Links;
import ch.rmy.android.statusbar_tacho.utils.Settings;
import ch.rmy.android.statusbar_tacho.utils.SimpleItemSelectedListener;

public class SettingsActivity extends BaseActivity {

    @Bind(R.id.speed)
    TextView speedView;
    @Bind(R.id.toggleButton)
    ToggleButton toggleButton;
    @Bind(R.id.unitSpinner)
    Spinner spinner;

    private Settings settings;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_settings);

        settings = new Settings(getContext());

        TachoService.setRunningState(this, TachoService.isRunning(this));

        toggleButton.setOnCheckedChangeListener(new OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                TachoService.setRunningState(getContext(), isChecked);
            }
        });

        List<CharSequence> list = new ArrayList<>();
        for (Unit unit : Units.UNITS) {
            list.add(getText(unit.getNameRes()));
        }
        ArrayAdapter<CharSequence> dataAdapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_item, list);
        dataAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinner.setAdapter(dataAdapter);
        spinner.setOnItemSelectedListener(new SimpleItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                settings.setUnit(Units.UNITS[position]);

                if (TachoService.isRunning(getContext())) {
                    TachoService.setRunningState(getContext(), false);
                    TachoService.setRunningState(getContext(), true);
                }
            }
        });
    }

    @Override
    int getNavigateUpIcon() {
        return 0;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.settings_activity_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_github: {
                Links.openGithub(getContext());
                return true;
            }
            default:
                return super.onOptionsItemSelected(item);
        }
    }
}

File: StatusBarSpeedometer/app/src/main/java/ch/rmy/android/statusbar_tacho/activities/SettingsActivity.kt
package ch.rmy.android.statusbar_tacho.activities

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.WindowManager
import android.widget.ArrayAdapter
import ch.rmy.android.statusbar_tacho.R
import ch.rmy.android.statusbar_tacho.location.SpeedWatcher
import ch.rmy.android.statusbar_tacho.services.SpeedometerService
import ch.rmy.android.statusbar_tacho.units.Unit
import ch.rmy.android.statusbar_tacho.units.Units
import ch.rmy.android.statusbar_tacho.utils.*
import kotlinx.android.synthetic.main.activity_settings.*

class SettingsActivity : BaseActivity() {

    override val navigateUpIcon = 0

    private var permissionManager: PermissionManager? = null
    private var speedWatcher: SpeedWatcher? = null
    private var unit: Unit? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_settings)

        val settings = Settings(context)
        unit = settings.unit
        onUnitChanged()
        setupUnitSelector(settings)

        speedWatcher = destroyer.own(SpeedWatcher(context))
        permissionManager = PermissionManager(context)

        toggleButton.setOnCheckedChangeListener { _, isChecked -> toggleState(isChecked) }
        speedWatcher!!.speedSource.bind(object : EventSource.Observer<Float?> {
            override fun on(currentSpeed: Float?) {
                if (currentSpeed == null) {
                    updateSpeedViews(0f)
                } else {
                    updateSpeedViews(currentSpeed)
                }
            }
        }, speedWatcher!!.currentSpeed)

        Dialogs.showIntroMessage(context, settings)
    }

    private fun updateSpeedViews(speed: Float) {
        val convertedSpeed = unit!!.convertSpeed(speed)
        speedGauge.value = convertedSpeed
        speedText.text = if (!toggleButton.isChecked) {
            getString(R.string.idle_speed)
        } else if (!speedWatcher!!.isGPSEnabled) {
            getString(R.string.gps_disabled)
        } else if (!speedWatcher!!.hasLocationPermission()) {
            getString(R.string.permission_missing)
        } else {
            getString(
                    R.string.speed_format_without_unit,
                    convertedSpeed
            )
        }
    }

    private fun onUnitChanged() {
        speedGauge.maxValue = unit!!.maxValue.toFloat()
        speedGauge.markCount = unit!!.steps + 1
    }

    override fun onStart() {
        super.onStart()
        initState()
    }

    private fun initState() {
        val state = SpeedometerService.isRunning(context)
        SpeedometerService.setRunningState(context, state)
        toggleButton.isChecked = SpeedometerService.isRunning(context)
        speedWatcher!!.toggle(state)
    }

    override fun onStop() {
        super.onStop()
        speedWatcher!!.disable()
    }

    private fun toggleState(state: Boolean) {
        if (state && !permissionManager!!.hasLocationPermission()) {
            toggleButton.isChecked = false
            permissionManager!!.requestLocationPermission(this)
            return
        }

        if (state) {
            window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        } else {
            window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        }

        speedWatcher!!.toggle(state)
        SpeedometerService.setRunningState(context, state)
        updateSpeedViews(0f)
    }

    private fun setupUnitSelector(settings: Settings) {
        val unitNames = Units.UNITS.map { getText(it.nameRes) }
        val dataAdapter = ArrayAdapter(context, android.R.layout.simple_spinner_item, unitNames)
        dataAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        unitSpinner.adapter = dataAdapter
        unitSpinner.onItemSelectedListener = object : SimpleItemSelectedListener() {
            override fun onItemSelected(position: Int) {
                unit = Units.UNITS[position]
                settings.unit = unit as Unit
                onUnitChanged()

                if (SpeedometerService.isRunning(context)) {
                    SpeedometerService.restart(context)
                }
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.settings_activity_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_github -> {
                Links.openGithub(context)
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        if (permissionManager!!.wasGranted(grantResults)) {
            toggleButton.isChecked = true
        }
    }

}
